### ISSUE TITLE
Semantic Tokens: Streaming and refinement

### ISSUE THREAD

**Issue by LukaszMendakiewicz** [microsoft/vscode#87934]
Assuming that streaming for semantic tokens will be offered in its LSP level specification and its implementation will be QoI in editors implementing the spec.

Below is how the colorization appears in a large (20k lines) C++ file in VS today (the animation shows the map-mode scrollbar which is providing the overview over the editor colors):

<details><summary>EXPAND TO SEE ANIMATION</summary>

![color](https://user-images.githubusercontent.com/6934054/71627311-97d36300-2ba6-11ea-9cb7-f8b126f63cc9.gif)

</details>

There are three relevant phases visible:

  1. 0-2 s: the text with syntactic colorization is being rendered
  2. 4.2 s: some region graying appears
  3. 4.6-11 s: full semantic colorization appears from top to bottom of the file segment by segment

We can suppose syntactic/first-pass colorization will be happening most commonly on the editor side (e.g. with TextMate). However if some LSP server would like to have its own quick-pass colorization shown before doing the heavier semantic analysis, having the ability to refine/push updated results might be needed.

Region graying for the whole file might be folded into phase 3 without much experience loss in this case, but perhaps there are LSP servers that would like to have similar information sent early still.

Full semantic colorization is done in chunks as you can see in the animation. The entire process takes ~7 seconds in this example file.

This raises some points:

1) While the results for the whole file are being computed (irrespective of streaming or not) in that 7 second period and the user moves their viewport to a different section of the file, would the editor/client be issuing another request for the new visible range? How would the delta results be reconciled?

2) Similar thing when the edits are happening while the token classification request is pending, would the editor be eagerly requerying with multiple requests in flight, or would it be waiting for the prior requests to finish? If the former, the same question about the delta results reconciliation.

**Comment by dbaeumer** [comment]
@alexdima FYI.

**Comment by alexdima** [comment]
@LukaszMendakiewicz Thank you for the great questions. *Please note that what I am describing here is not 100% implemented, but just shows what my thinking is*.

Since you have written your issue, the API now separates things and there are two distinct semantic tokens providers, a `DocumentSemanticTokensProvider` and a `DocumentRangeSemanticTokensProvider`. The `DocumentRangeSemanticTokensProvider` will be called with the viewport and will not support deltas. A language implementation could implement both or just one of them.

### 1. How things work until the document request finishes.

* When the user first opens a file, the editor will first and foremost make a request to the `DocumentRangeSemanticTokensProvider` with the current viewport. Immediately, without waiting for the result, it will make a request to the `DocumentSemanticTokensProvider`. So there will be two requests running at the same time, the request for the viewport and the request for the entire document. (*Certain language servers are single threaded and might see this as one request at a time, but on our side we will actually run them in parallel so servers that support multithreading could run them in parallel*).
* If the user just sits idle, when the viewport request resolves, the semantic tokens of the viewport will be rendered. When the entire document request resolves, the semantic tokens of the viewport will be dropped and the entire document result will be used. Once the entire document request resolves, the range based provider will never be invoked again for the document. The entire document tokens will be used for rendering and when the document gets edited, the tokens will be kept up-to-date via the delta mechanism.
* If the user scrolls in the meantime, subsequent viewport requests will be made and the cancellation token will be fired on the previous ones. Even if the user scrolls just a few lines, there is no delta story. My thinking is that the viewport has a limited number of semantic tokens and a delta story would only complicate things needlessly. 

### 2. Editing and the full document request

* If the editor has a full document request running, and the user types:
  * the current request is not canceled
  * there is no other request made
  * the edit is recorded locally in a queue

* When the document request resolves:
  * the tokens are "reconciled" with the queued edits and then applied to the text buffer
  * a new delta request is made with the previous result id

---

From what I described above, the part about the viewport provider is not yet implemented.

Regarding streaming, IMHO it would only make sense for the full document provider, but given the viewport provider can "cover things" until the full document provider resolves, I believe things might just work even without streaming...
